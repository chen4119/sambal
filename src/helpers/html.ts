import {Parser} from "htmlparser2";
import {Handler} from "htmlparser2/lib/Parser";

// TODO: spacing not used.  Remove?
const SPACING = 4;
const SAMBAL_SLOT_NAME = "sambal-slot";
const SELF_CLOSING_TAGS = [
    'area',
    'base',
    'br',
    'embed',
    'hr',
    'iframe',
    'img',
    'input',
    'link',
    'meta',
    'param',
    'source',
    'track',
    SAMBAL_SLOT_NAME
];

type HtmlAttributes = {
    [propName: string]: string
};

export async function replaceScriptSrc(srcHtml: string, replace: (src: string) => string) {
    return new Promise<string>((resolve, reject) => {
        const handler = getHtmlSerializer(resolve, reject, null, (name, attribs) => {
            if (name === "script" && attribs.src) {
                return {
                    ...attribs,
                    src: replace(attribs.src)
                };
            }
            return attribs;
        });
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

export async function hydrateSambalSlots(
    srcHtml: string,
    render: (id: string) => Promise<string>) {

    const slots = await searchForSambalSlots(srcHtml);
    for (const slot of slots) {
        slot.html = await render(slot.id);
    }

    return new Promise<string>((resolve, reject) => {
        let slotIndex = -1;
        const handler = getHtmlSerializer(resolve, reject, (name, attribs) => {
            if (name === SAMBAL_SLOT_NAME) {
                slotIndex++;
                return slots[slotIndex].html ? slots[slotIndex].html : "";
            }
            return null;
        }, null);
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

function getHtmlSerializer(
    resolve,
    reject,
    replace: (name: string, attribs: HtmlAttributes) => string,
    editAttribs: (name: string, attribs: HtmlAttributes) => HtmlAttributes) {
    const stack: string[] = [];
    let html = '';
    let deleteTag: string = null; 
    const handler: Partial<Handler> = {
        onopentag: (name,  attribs) => {
            stack.unshift(name);
            if (deleteTag != null) {
                return;
            }
            const replaceHtml = replace ? replace(name, attribs) : null;
            if (replaceHtml !== null) {
                deleteTag = name;
                html += replaceHtml;
            } else {
                const updatedAttribs = editAttribs ? editAttribs(name, attribs) : attribs;
                html += addOpeningTag(0, name, updatedAttribs);
            }
            
        },
        ontext:(data) => {
            if (deleteTag !== null) {
                return;
            }
            html += data;
        },
        onclosetag: (name) => {
            const tagName = stack.shift();
            if (deleteTag === tagName) {
                deleteTag = null;
            } else {
                html += addClosingTag(0, name);
            }
        },
        onprocessinginstruction: (name, data) => {
            html += addProcessingInstruction(0, data);
        },
        onerror: (err) => {
            reject(err);
        },
        onend: () => {
            resolve(html);
        }
    };
    return handler;
}

async function searchForSambalSlots(srcHtml: string) {
    return new Promise<any[]>((resolve, reject) => {
        const slots = [];
        const handler: Partial<Handler> = {
            onopentag: (name,  attribs) => {
                if (name === SAMBAL_SLOT_NAME) {
                    slots.push({
                        id: attribs.link
                    });
                }
            },
            onerror: (err) => {
                reject(err);
            },
            onend: () => {
                resolve(slots);
            }
        };
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

function addOpeningTag(indent: number, name: string, attribs?: HtmlAttributes) {
    let attributes = '';
    if (attribs) {
        for (const key of Object.keys(attribs)) {
            attributes += ` ${key}="${attribs[key]}"`;
        }
    }
    return `${makeIndent(indent * SPACING)}<${name}${attributes}>`;
}

function addClosingTag(indent: number, name: string) {
    if (SELF_CLOSING_TAGS.indexOf(name) >= 0) {
        return '';
    }
    return `${makeIndent(indent * SPACING)}</${name}>`;
}

function addProcessingInstruction(indent: number, data: string) {
    return `${makeIndent(indent * SPACING)}<${data}>\n`;
}

function makeIndent(numSpaces: number) {
    let indent = '';
    for (let i = 0; i < numSpaces; i++) {
        indent += ' ';
    }
    return indent;
}
