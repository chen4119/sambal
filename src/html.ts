import {Parser} from "htmlparser2";
import {Handler} from "htmlparser2/lib/Parser";
import {safeParseJson} from "./utils";
const SPACING: number = 4;
const SELF_CLOSING_TAGS = [
    'area',
    'base',
    'br',
    'embed',
    'hr',
    'iframe',
    'img',
    'input',
    'link',
    'meta',
    'param',
    'source',
    'track'
];
const OPEN = "open";
const TEXT = "text";
const CLOSE = "close";

export type HtmlAttributes = {
    [propName: string]: string
};

export type HtmlNode = {
    name: string,
    attributes?: HtmlAttributes
    body?: string,
}

export async function getJsonLd(srcHtml: string) {
    return new Promise<any[]>((resolve, reject) => {
        let allJsonLds = [];
        let isJsonLdInBody = false;
        const handler: Partial<Handler> = {
            onopentag: (name,  attribs) => {
                if (name === 'script' && attribs.type === 'application/ld+json') {
                    isJsonLdInBody = true;
                }
            },
            ontext:(data) => {
                if (isJsonLdInBody) {
                    const jsonld = safeParseJson(data);
                    if (jsonld) {
                        allJsonLds.push(jsonld);
                    }
                    isJsonLdInBody = false;
                }
            },
            onerror: (err) => {
                reject(err);
            },
            onend: () => {
                resolve(allJsonLds);
            }
        };
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

export async function editHtml(
    srcHtml: string,
    editAttribs: (name: string, attribs: HtmlAttributes) => HtmlAttributes
) {
    return new Promise<string>((resolve, reject) => {
        let html: string = '';
        const handler: Partial<Handler> = {
            onopentag: (name,  attribs) => {
                const updatedAttribs = editAttribs(name, attribs);
                html += addOpeningTag(0, name, updatedAttribs);
            },
            ontext:(data) => {
                html += data;
            },
            onclosetag: (name) => {
                html += addClosingTag(0, name);
            },
            onprocessinginstruction: (name, data) => {
                html += addProcessingInstruction(0, data);
            },
            onerror: (err) => {
                reject(err);
            },
            onend: () => {
                resolve(html);
            }
        };
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

export async function prettify(srcHtml: string, appendHeadNodes: HtmlNode[] = []) {
    return new Promise<string>((resolve, reject) => {
        let foundHead = false;
        let isAppendedHeadNodes = appendHeadNodes.length > 0 ? false : true;
        const preOpenTag = (name) => {
            if (name === 'head') {
                foundHead = true;
            } else if (name === 'body' && !foundHead) {
                // automatically add head if none found by the time we get to <body>
                handler.onopentag('head', null);
                handler.onclosetag('head');
            }
        };
        const preCloseTag = (name) => {
            if (name === 'head' && !isAppendedHeadNodes) {
                for (const node of appendHeadNodes) {
                    handler.onopentag(node.name, node.attributes);
                    handler.ontext(node.body);
                    handler.onclosetag(node.name);
                }
                isAppendedHeadNodes = true;
            }
        };
        const handler = getHtmlSerializer(resolve, reject, preOpenTag, preCloseTag);
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

function getHtmlSerializer(resolve, reject, preOpenTag?, preCloseTag?) {
    let html: string = '';
    let indent: number = 0;
    let prev: string = null;
    let insidePreTag: boolean = false;
    const handler: Partial<Handler> = {
        onopentag: (name,  attribs) => {
            if (preOpenTag) {
                preOpenTag(name, attribs);
            }
            if (!insidePreTag && prev === OPEN) {
                indent ++;
                html += `\n${addOpeningTag(indent, name, attribs)}`;
            } else if (!insidePreTag && prev === CLOSE) {
                html += addOpeningTag(indent, name, attribs);
            } else {
                html += addOpeningTag(0, name, attribs);
            }
            prev = OPEN;
            if (name === 'pre') {
                insidePreTag = true;
            }
        },
        ontext:(data) => {
            const trimmed = data.trim();
            if (trimmed) {
                html += trimmed;
                prev = TEXT;
            }
        },
        onclosetag: (name) => {
            if (preCloseTag) {
                preCloseTag(name);
            }
            if (insidePreTag) {
                html += addClosingTag(0, name);
            } else if (prev === TEXT || prev === OPEN) {
                html += `${addClosingTag(0, name)}\n`;
            } else {
                indent = Math.max(0, indent - 1);
                html += `${addClosingTag(indent, name)}\n`;
            }
            prev = CLOSE;
            if (name === 'pre') {
                insidePreTag = false;
                html += '\n';
            }
        },
        onprocessinginstruction: (name, data) => {
            html += addProcessingInstruction(indent, data);
        },
        onerror: (err) => {
            reject(err);
        },
        onend: () => {
            resolve(html);
        }
    };
    return handler;
}

function addOpeningTag(indent: number, name: string, attribs?: HtmlAttributes) {
    let attributes = '';
    if (attribs) {
        for (const key of Object.keys(attribs)) {
            attributes += ` ${key}="${attribs[key]}"`;
        }
    }
    return `${makeIndent(indent * SPACING)}<${name}${attributes}>`;
}

function addClosingTag(indent: number, name: string) {
    if (SELF_CLOSING_TAGS.indexOf(name) >= 0) {
        return '';
    }
    return `${makeIndent(indent * SPACING)}</${name}>`;
}

function addProcessingInstruction(indent: number, data: string) {
    return `${makeIndent(indent * SPACING)}<${data}>\n`;
}

function makeIndent(numSpaces: number) {
    let indent = '';
    for (let i = 0; i < numSpaces; i++) {
        indent += ' ';
    }
    return indent;
}
