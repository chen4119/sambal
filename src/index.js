export {SambalApp} from "./app";
export {store} from "./store";