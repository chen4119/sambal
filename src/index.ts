import SambalCollection from "./SambalCollection";
import Logger from "./Logger";

export {
    toJsonLdGraph,
    toSchemaOrgJsonLd,
    graphToCondensedJsonLd,
    hydrateJsonLd
} from "sambal-jsonld";

export {SambalCollection};
export {Logger};
export {template} from "./template";
export {loadContent} from "./utils";
export {render} from "./operators/render";
export {aggregateBy} from "./operators/aggregateBy";
export {loadJsonLd} from "./operators/loadJsonLd";
export {paginate} from "./operators/paginate";
export {toHtml} from "./operators/toHtml";
export {listJsonLd} from "./operators/listJsonLd";
export {pushJsonLd} from "./operators/pushJsonLd";