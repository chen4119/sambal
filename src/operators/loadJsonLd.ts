import {Observable, pipe} from "rxjs";
import {mergeMap} from "rxjs/operators";
import {loadContent} from "../utils";
import {hydrateJsonLd} from "sambal-jsonld";

export function loadJsonLd(hydrateOptions?: {
    fetcher?: (url: string) => Promise<any>,
    context?: any
}) {
    return pipe<Observable<string>, Observable<any>>(
        mergeMap(async src => {
            const content = await loadContent(src);
            if (hydrateOptions) {
                return await hydrateJsonLd(content, hydrateOptions.fetcher, hydrateOptions.context);
            }
            return content;
        })
    );
}