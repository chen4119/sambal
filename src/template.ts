
export async function template(literals, ...expressions) {
    const resolvedExpr = await Promise.all(expressions);
    let string = "";
    for (const [i, val] of resolvedExpr.entries()) {
        if (Array.isArray(val)) {
            const resolvedValues = await Promise.all(val);
            string += `${literals[i]}${resolvedValues.filter(v => v !== null && v !== undefined).join("")}`;
        } else {
            string += `${literals[i]}${(val === null || val === undefined) ? "" : val}`;
        }
    }
    string += literals[literals.length - 1];
    return string;
}
