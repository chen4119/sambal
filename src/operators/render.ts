import {pipe, Observable} from "rxjs";
import {mergeMap} from "rxjs/operators";
import LocalCss from "../LocalCss";
import {SambalData, SAMBAL_INTERNAL} from "../constants";

export function render(renderer: (props: any) => Promise<string>) {
    return pipe<Observable<any>, Observable<SambalData>>(
        mergeMap(async (data: any) => {
            const css = new LocalCss();

            if (!data[SAMBAL_INTERNAL]) {
                data[SAMBAL_INTERNAL] = {};
            }
            data[SAMBAL_INTERNAL].css = css;
            data[SAMBAL_INTERNAL].html = await renderer({...data, css: css});
            return data;
        })
    );
}
