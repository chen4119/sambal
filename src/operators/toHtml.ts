import {pipe} from "rxjs";
import {mergeMap} from "rxjs/operators";
import {SAMBAL_INTERNAL} from "../constants";
import Logger from "../Logger";
import {prettify, HtmlNode, HtmlAttributes} from "../html";
import {toJsonLdGraph, SCHEMA_CONTEXT} from "sambal-jsonld";


const log: Logger = new Logger({name: "toHtml"});
export function toHtml(mutateHtml: {
    editAttribs?: (name: string, attribs: HtmlAttributes) => HtmlAttributes,
    prepend?: (name: string, attribs: HtmlAttributes) => HtmlNode[],
    append?: (name: string, attribs: HtmlAttributes) => HtmlNode[],
    replace?: (name: string, attribs: HtmlAttributes) => HtmlNode
} = {}) {
    return pipe(
        mergeMap(async d => {
            if (d[SAMBAL_INTERNAL] && d[SAMBAL_INTERNAL].html) {
                const injectHeadNodes: HtmlNode[] = [];
                if (d[SAMBAL_INTERNAL].css && d[SAMBAL_INTERNAL].css.hasSheets()) {
                    injectHeadNodes.push({
                        name: "style",
                        body: d[SAMBAL_INTERNAL].css.getCss()
                    });
                }
                if (d[SAMBAL_INTERNAL].jsonld && d[SAMBAL_INTERNAL].jsonld.length > 0) {
                    const schemaOrgJson = toJsonLdGraph(d[SAMBAL_INTERNAL].jsonld, SCHEMA_CONTEXT);
                    injectHeadNodes.push({
                        name: "script",
                        body: JSON.stringify(schemaOrgJson),
                        attributes: {
                            type: "application/ld+json"
                        }
                    });
                }
                return await editHtml(
                    d[SAMBAL_INTERNAL].html,
                    injectHeadNodes,
                    mutateHtml.editAttribs ? mutateHtml.editAttribs : NOOP_EDIT_ATTRIBS,
                    mutateHtml.prepend ? mutateHtml.prepend : NOOP_PREPEND,
                    mutateHtml.append ? mutateHtml.append : NOOP_APPEND,
                    mutateHtml.replace ? mutateHtml.replace : NOOP_REPLACE,
                );
            }
            log.warn("No html rendered.  Need to call render first");
            return "";
        })
    );
}

async function editHtml(srcHtml: string,
    injectHeadNodes: HtmlNode[],
    userDefinedEditAttribs: (name: string, attribs: HtmlAttributes) => HtmlAttributes,
    userDefinedPrepend: (name: string, attribs: HtmlAttributes) => HtmlNode[],
    userDefinedAppend: (name: string, attribs: HtmlAttributes) => HtmlNode[],
    userDefinedReplace: (name: string, attribs: HtmlAttributes) => HtmlNode
) {
    let isAdded = injectHeadNodes.length > 0 ? false : true;
    return await prettify(
        srcHtml,
        userDefinedEditAttribs,
        (name: string, attribs: HtmlAttributes) => {
            let prependNodes = userDefinedPrepend(name, attribs);
            if (!isAdded && name === 'body') {
                isAdded = true;
                const node = {
                    name: 'head',
                    body: injectHeadNodes
                };
                if (!prependNodes) {
                    prependNodes = [node];
                } else {
                    prependNodes.push(node);
                }
            }
            return prependNodes;
        },
        (name: string, attribs: HtmlAttributes) => {
            let appendNodes = userDefinedAppend(name, attribs);
            if (!isAdded && name === 'head') {
                isAdded = true;
                if (!appendNodes) {
                    appendNodes = injectHeadNodes;
                } else {
                    appendNodes = appendNodes.concat(injectHeadNodes);
                }
            }
            return appendNodes;
        },
        userDefinedReplace
    );
}

const NOOP_EDIT_ATTRIBS = (name: string, attribs: HtmlAttributes) => attribs;
const NOOP_PREPEND = (name: string, attribs: HtmlAttributes) => null;
const NOOP_APPEND = (name: string, attribs: HtmlAttributes) => null;
const NOOP_REPLACE = (name: string, attribs: HtmlAttributes) => null;