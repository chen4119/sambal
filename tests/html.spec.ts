import {from} from "rxjs";
import {template} from "../src/template";
import {loadJsonLd} from "../src/operators/loadJsonLd";
import {render} from "../src/operators/render";
import {toHtml} from "../src/operators/toHtml";
import {prettify, HtmlNode, HtmlAttributes} from "../src/html";

const NOOP_EDIT_ATTRIBS = (name: string, attribs: HtmlAttributes) => attribs;
const NOOP_PREPEND = (name: string, attribs: HtmlAttributes) => null;
const NOOP_APPEND = (name: string, attribs: HtmlAttributes) => null;
const NOOP_REPLACE = (name: string, attribs: HtmlAttributes) => null;

describe('html', () => {
    const html = `
        <!doctype html>
        <html>
            <head>
                <meta charset="UTF-8">
                <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
                <meta http-equiv="X-UA-Compatible" content="ie=edge">
                <meta http-equiv="Cache-Control" content="no-cache, no-store, must-revalidate">
                <base href="/">
                <script src="./js/index.js"></script>
            </head>
            <body>
                <main>
                    <h1>hello world</h1>

                    <p>My paragraph</p>
                </main>
                <a href="/hfoi">link</a>
                <img href="/image.png">
                <br>

                <br>
            </body>
        </html>
    `;

    it('prettify', async () => {
        const result = await prettify(html, NOOP_EDIT_ATTRIBS, NOOP_PREPEND, NOOP_APPEND, NOOP_REPLACE);
        expect(result).toMatchSnapshot();
    });

    it('change src attrib', async () => {
        const result = await prettify(html, (name, attribs) => {
            if (name === 'script') {
                return {
                    src: 'https://localhost:3000'
                }
            }
            return attribs;
        }, NOOP_PREPEND, NOOP_APPEND, NOOP_REPLACE);
        expect(result).toMatchSnapshot();
    });

    it('replace node', async () => {
        const result = await prettify(html, NOOP_EDIT_ATTRIBS, NOOP_PREPEND, NOOP_APPEND, (name, attribs) => {
            if (name === 'img') {
                return {
                    name: 'picture',
                    body: [
                        {name: 'source', attributes: {media: "(min-width: 480px)", srcset: "img1.jpg"}},
                        {name: 'source', attributes: {media: "(min-width: 650px)", srcset: "img2.jpg"}}
                    ]
                };
            }
            return null;
        });
        expect(result).toMatchSnapshot();
    });

    it('preserve <pre>', async () => {
        const result = await from(['tests/post.md'])
        .pipe(loadJsonLd())
        .pipe(render(({text}) => {
            return template`
                <!doctype html>
                <html>
                    <body>
                        ${template`
                            <main>
                                ${text}
                            </main>
                        `}
                    </body>
                </html>
            `;
        }))
        .pipe(toHtml())
        .toPromise();
        expect(result).toMatchSnapshot();
    });
})