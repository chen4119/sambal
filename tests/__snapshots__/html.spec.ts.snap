// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html change src attrib 1`] = `
"<!doctype html>
<html>
    <head>
        <meta charset=\\"UTF-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\">
        <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
        <meta http-equiv=\\"Cache-Control\\" content=\\"no-cache, no-store, must-revalidate\\">
        <base href=\\"/\\">
        <script src=\\"https://localhost:3000\\"></script>
    </head>
    <body>
        <main>
            <h1>hello world</h1>
            <p>My paragraph</p>
        </main>
        <a href=\\"/hfoi\\">link</a>
        <img href=\\"/image.png\\">
        <br>
        <br>
    </body>
</html>"
`;

exports[`html preserve <pre> 1`] = `
"<!doctype html>
<html>
    <body>
        <main>
            <p>This is my very first post1</p>
            <pre><code class=\\"language-ShellSession\\">$ npm install --save-dev sambal-cli</code></pre>

            <p>Sambal has <a href=\\"https://webpack.org\\">webpack</a>, <a href=\\"https://webpack.org\\">webpack</a>, <a href=\\"https://webpack.org\\">webpack</a> to do with something else</p>
            <pre><code class=\\"language-js\\">import {pipe, from, of} from &quot;rxjs&quot;;
import {mergeMap, map, groupBy, reduce, toArray} from &quot;rxjs/operators&quot;;

function aggregateBy(field) {
    return pipe(
        map((data) =&gt; data[field]),
        mergeMap(fieldValue =&gt; Array.isArray(fieldValue) ? from(fieldValue) : of(fieldValue)),
        groupBy(v =&gt; v),
        mergeMap(group$ =&gt;
            group$.pipe(reduce((acc, cur) =&gt; {
                acc.count++;
                return acc;
            }, {key: group$.key, count: 0}))    
        ),
        toArray()
    );
}</code></pre>

        </main>
    </body>
</html>"
`;

exports[`html prettify 1`] = `
"<!doctype html>
<html>
    <head>
        <meta charset=\\"UTF-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\">
        <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
        <meta http-equiv=\\"Cache-Control\\" content=\\"no-cache, no-store, must-revalidate\\">
        <base href=\\"/\\">
        <script src=\\"./js/index.js\\"></script>
    </head>
    <body>
        <main>
            <h1>hello world</h1>
            <p>My paragraph</p>
        </main>
        <a href=\\"/hfoi\\">link</a>
        <img href=\\"/image.png\\">
        <br>
        <br>
    </body>
</html>"
`;

exports[`html replace node 1`] = `
"<!doctype html>
<html>
    <head>
        <meta charset=\\"UTF-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\">
        <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
        <meta http-equiv=\\"Cache-Control\\" content=\\"no-cache, no-store, must-revalidate\\">
        <base href=\\"/\\">
        <script src=\\"./js/index.js\\"></script>
    </head>
    <body>
        <main>
            <h1>hello world</h1>
            <p>My paragraph</p>
        </main>
        <a href=\\"/hfoi\\">link</a>
        <picture>
            <source media=\\"(min-width: 480px)\\" srcset=\\"img1.jpg\\">
            <source media=\\"(min-width: 650px)\\" srcset=\\"img2.jpg\\">
        </picture>
        <br>
        <br>
    </body>
</html>"
`;
